### PR TITLE
Feat: Player Rating Endpoints

### PR DESCRIPTION
**Pull request by invaderb**
feat: add ratings endpoints
feat: add scheduled job
test: add tests for rating generation
chore: add endpoints to readme

Adds the following endpoints to the service:
/v1/ratings
/v1/ratings/{player_name}

Changes:
Adds the following dependency @nest/schedule

### CODE REVIEW

File: src/app.module.ts [hunk 1]
```diff
@@ -1,29 +1,42 @@
...
 		ConfigModule.forRoot(),
 		TypeOrmModule.forRoot({
+			name: 'games',
 			type: 'sqlite',
 			database: process.env.DB_PATH + 'games.db',
 			entities: [__dirname + '/**/*.entity{.ts,.js}'],
 			synchronize: false,
 		}),
+		TypeOrmModule.forRoot({
+			name: 'default',
```
Comment (nitzel) [inline src/app.module.ts:24]: `BaseDataSourceOptions.name` is [deprecated](https://github.com/typeorm/typeorm/blob/d8a2e3730f12bb2b8e521635e176a284594121f3/src/data-source/BaseDataSourceOptions.ts#L23), [explanation:  `we don't need names anymore since we are going to drop all related methods relying on this property.`](https://github.com/typeorm/typeorm/blob/d8a2e3730f12bb2b8e521635e176a284594121f3/src/data-source/DataSource.ts#L62)

The alternative seems to be to tell TypeOrm here which entities are related to which database:

https://docs.nestjs.com/techniques/database#repository-pattern

```ts
TypeOrmModule.forRoot({ 
      // ...
      database: process.env.DB_PATH + 'players.db',
      entities: [Ratings, Players],
    }),
```

File: src/models/ratings/ratings.module.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { RatingService } from './ratings.service';
+import { Ratings } from './entities/ratings.entity';
+import { RatingTask } from './tasks/rating.task';
+import { Players } from './entities/players.entity';
+import { Games } from '../games-history/entities/games.entity';
+
+@Module({
+	controllers: [RatingsController],
+	imports: [
+		TypeOrmModule.forFeature([Ratings, Players], 'default'),
```
Comment (nitzel) [inline src/models/ratings/ratings.module.ts:15]: no need to specify connection name, see other comment. `[Ratings, Players, Games]` probably can be imported in a single `forFeature` call.

File: src/models/ratings/ratings.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+		}).compile();
+
+		service = module.get<RatingService>(RatingService);
+	});
+
+	it('should be defined', () => {
+		expect(service).toBeDefined();
+	});
+	
+	// describe('Generate Ratings', () => {
```
Comment (nitzel) [inline src/models/ratings/ratings.service.spec.ts]: ?

File: src/models/ratings/ratings.service.ts [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+			});
+			return {
+				items: results[0] || [],
+				total: results[1] || 0,
+				page: page + 1,
+				perPage: limit,
+				totalPages: Math.ceil(results[1] / limit),
+			};
+		} catch (error) {
+			console.error(error);
```
Comment (nitzel) [inline src/models/ratings/ratings.service.ts]: why not use `this.logger.error`?

File: .env [hunk 5]
```diff
@@ -5,4 +5,5 @@ NODE_ENV=local
 APP_ENV=Local
 CORS_DOMAIN=*
 ANON_DB_PATH=/Users/braydonharris/dev/personal-projects/tak/server/games_anon.db
-DIST_PATH=dist/
\ No newline at end of file
+DIST_PATH=dist/
+PREVIOUS_FILE=previous.txt
```
Comment (nitzel) [inline .env:9]: What do you think about renaming this to `last_used_game_id.txt` or similar?

**Review by nitzel — COMMENTED** [review]
LGTM - just a few minor suggestions.

Should this be pointing to the `dev` branch?

File: src/models/ratings/ratings.service.ts (hunk 4 above)
Comment (invaderb) [reply src/models/ratings/ratings.service.ts]: Just forgot to move it over to logger, which I don't mind using but it does strip out the line number where the error happened which makes it harder to debug...

File: .env (hunk 5 above)
Comment (invaderb) [reply .env:9]: I'm fine either way this was just copied over from the old script.

File: src/app.module.ts (hunk 1 above)
Comment (invaderb) [reply src/app.module.ts:24]: Interesting I wonder how it's going to work with two different db's then; I'll look into it and push a fix.

File: src/models/ratings/ratings.service.spec.ts (hunk 3 above)
Comment (invaderb) [reply src/models/ratings/ratings.service.spec.ts]: more so just left this in here to come back and add more unit tests for the generateRating method but I'll just go ahead and remove it, as the main business logic has testing.

File: src/app.module.ts (hunk 1 above)
Comment (invaderb) [reply src/app.module.ts:24]: Yeah it breaks if you remove the name when managing multiple databases on both the for root and for feature.  it needs some way to differentiate between the 2 and I found that you need to have a default for the datasource to fall back on. if we had one db this wouldn't be a problem and could use this pattern.

File: .env (hunk 5 above)
Comment (invaderb) [reply .env:9]: for backwards compatibility going to leave this as is so there is parity between what's used on the server is the same as in the repo.

File: src/app.module.ts (hunk 1 above)
Comment (invaderb) [reply src/app.module.ts:24]: Screenshot 2023-05-06 at 3 18 12 PM

Comment (invaderb) [reply src/app.module.ts:24]: I also tried the autoLoadEntities and that didn't work either.

Comment (invaderb) [reply src/app.module.ts:24]: Ah yeah here in the docs, https://docs.nestjs.com/techniques/database#multiple-databases
```
In this case, data source naming becomes mandatory.
```

Comment (nitzel) [reply src/app.module.ts:24]: Ugh sorry for sending you down that path. I'm confused why they a) deprecated `name` if it's still necessary and b) what is `entities` for?

File: src/models/ratings/dto/ratings.dto.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	@ApiProperty()
+		totalPages: number;
+}
+
+export class RatingQuery {
+	@ApiProperty({
+		description: 'Number of results per page',
+		required: false
+	})
+		limit?: string;
```
Comment (nitzel) [inline src/models/ratings/dto/ratings.dto.ts]: Should this be a `number`? Same with `RatingQuery .page`?

File: src/models/ratings/dto/ratings.dto.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	@ApiProperty({
+		description: 'page number',
+		required: false
+	})
+		page?: string;
+	@ApiProperty({
+		description: 'Optional skip if you dont use page',
```
Comment (nitzel) [inline src/models/ratings/dto/ratings.dto.ts]: what is `skip` for? What values can I pass in?

File: src/models/ratings/ratings.controller.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	}
+	
+	@ApiOperation({ operationId: 'Player Rating', summary: 'Get Players Rating' })
+	@ApiResponse({status: 200, type: RatingList, description: 'Returns list of games' })
+	@ApiResponse({status: 404, type: DefaultExceptionDto, description: 'Returns 404 server error'})
+	@ApiResponse({status: 429, type: DefaultExceptionDto, description: 'Returns 429 too many requests error'})
+	@ApiResponse({status: 500, type: DefaultExceptionDto, description: 'Returns 500 server error'})
+	@ApiParam({ name: 'name', description: 'player name', required: true })
+	@Get('/:name')
+	findRating(@Param() param: any): Promise<Rating> {
```
Comment (nitzel) [inline src/models/ratings/ratings.controller.ts]: I think this can be improved to 
```ts
	findRating(@Param('name') name: string): Promise<Rating> {
		return this.service.getPlayersRating(name);
	}
```

File: src/models/ratings/ratings.service.ts (hunk 4 above)
Comment (nitzel) [reply src/models/ratings/ratings.service.ts]: I think the `console.log` can go the next time you need to touch this file 😄

File: src/models/ratings/dto/ratings.dto.ts (hunk 7 above)
Comment (invaderb) [reply src/models/ratings/dto/ratings.dto.ts]: it's the total number of items to skip when doing pagination, can clean up this description

File: src/models/ratings/dto/ratings.dto.ts (hunk 6 above)
Comment (invaderb) [reply src/models/ratings/dto/ratings.dto.ts]: Since these come from the query string they are strings and not numbers and need to be parsed after the fact.

File: src/models/ratings/dto/ratings.dto.ts [hunk 9]
```diff
@@ -29,38 +29,38 @@ export class RatingList {
 
 export class RatingQuery {
 	@ApiProperty({
-		description: 'Number of results per page',
+		description: 'Number of results per pag for pagination. Optional.',
```
Comment (humanat) [inline src/models/ratings/dto/ratings.dto.ts]: pag = page

File: README.md [hunk 10]
```diff
@@ -53,6 +52,12 @@ https://api.{env}.playtak.com
 - /v1/games-historoy/ptn/{id}
 - /v1/hames-history/db
 
+- /v1/ratings
+- /v1/ratings{player_name}
+
+- /v1/ratings
+- /v1/ratings{player_name}
```
Comment (humanat) [inline README.md]: Should these have a / in them before the player_name or is that baked into the variable?

**Review by humanat — APPROVED** [review]
Approving, assuming that the comments are trivial to explain.

**Review by nitzel — APPROVED** [review]

File: README.md (hunk 10 above)
Comment (invaderb) [reply README.md]: yeah need to add the / I'll make an update to the readme and double check the endpoints

File: src/models/ratings/tasks/rating.task.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { Cron } from '@nestjs/schedule';
```
Comment (devp) [inline src/models/ratings/tasks/rating.task.ts:2]: TIL about node-cron. wow.

File: src/main.ts [hunk 12]
```diff
@@ -22,11 +23,6 @@ async function bootstrap() {
 	});
 	
 	const config = new DocumentBuilder()
-		.addApiKey({
```
Comment (devp) [inline src/main.ts:25]: Was this apiKey never used?

Comment (invaderb) [reply src/main.ts:25]: No this was just copy paste from the example

File: src/models/dto/response.dto.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable indent */
+import { ApiProperty } from '@nestjs/swagger';
+
+export class DefaultReposeDto {
```
Comment (devp) [inline src/models/dto/response.dto.ts]: typo

File: src/models/ratings/ratings.controller.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	@ApiResponse({status: 404, type: DefaultExceptionDto, description: 'Returns 404 server error'})
+	@ApiResponse({status: 429, type: DefaultExceptionDto, description: 'Returns 429 too many requests error'})
+	@ApiResponse({status: 500, type: DefaultExceptionDto, description: 'Returns 500 server error'})
+	@ApiParam({ name: 'name', description: 'player name', required: true })
+	@Get('/:name')
+	findRating(@Param('name') name: string): Promise<Rating | DefaultExceptionDto> {
+		return this.service.getPlayersRating(name);
+	}
+
+	// // manually kick off rating generation
```
Comment (devp) [inline src/models/ratings/ratings.controller.ts:36]: This is just a TODO but this should probably be a POST

File: src/models/ratings/tasks/rating.task.ts (hunk 11 above)
Comment (invaderb) [reply src/models/ratings/tasks/rating.task.ts:2]: yeah it's a nifty package and pretty easy to setup in nest.

File: src/models/ratings/ratings.service.ts [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+			gameRunner.release();
+			if (!has_error) {
+				this.logger.debug('Finished updating ratings');
+				await writeFile('previous.txt', lastUsedGame + ' ' + 'hash');
+			}
+		}
+	}
+
+	// returns the adjusted rating of a player
+	public adjustPlayer(
```
Comment (devp) [inline src/models/ratings/ratings.service.ts:353]: Note that this method and updateFatigue are mutations; they change the `player` object that is passed in. The description suggests otherwise. That may be fine though?

File: src/models/ratings/ratings.service.ts [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+		}
+		for (const a in player.fatigue) {
+			player.fatigue[a] *= MULTIPLIER;
+			if (a != opponentID && player.fatigue[a] < 0.01) {
+				delete player.fatigue[a];
+			}
+		}
+		return (player.fatigue[opponentID] = (player.fatigue[opponentID] || 0) + gameFactor);
+	}
+
```
Comment (devp) [inline src/models/ratings/ratings.service.ts:403]: contrasting to adjustPlayer, adjustedRating just returns the value

File: src/models/ratings/ratings.service.ts
Comment (devp) [inline src/models/ratings/ratings.service.ts:1]: Is the goal with ratings.service that the generateRating() code is ported from the nohat code mostly, and that we can run a command in NodeJS land rather than running a sciript?

generateRating() was too dense for much of a line-by-line reading -- its correctness will have to be proven out by running it over the real data and seeing that it's the same - but I don't know if it's worth rewriting in a future branch, either.

File: src/models/ratings/ratings.service.spec.ts
Comment (devp) [inline src/models/ratings/ratings.service.spec.ts:1]: I like the unit test, but have to vaguely trust what it is testing.

**Review by devp — APPROVED** [review]
LGTM but leaving some non-blocking comments for my understanding and some things I noticed. I most notably just skimmed over generateRating() because that was ported code that is hard to parse, and even riskier to rewrite without testing.

File: src/models/ratings/ratings.controller.ts (hunk 14 above)
Comment (invaderb) [reply src/models/ratings/ratings.controller.ts:36]: the endpoint isn't meant to be public, it's only here for testing locally why it's commented out the method type doesn't really matter in this case.

if we do ever make it public and put it behind some admin check we'll change it to a post at that point.

File: src/models/ratings/ratings.service.ts (hunk 16 above)
Comment (invaderb) [reply src/models/ratings/ratings.service.ts:403]: I updated the comments to explain it better what it's doing.

File: src/models/ratings/ratings.service.ts (hunk 15 above)
Comment (invaderb) [reply src/models/ratings/ratings.service.ts:353]: Updated the comments to explain what its doing it better.

File: src/models/ratings/ratings.service.ts
Comment (invaderb) [reply src/models/ratings/ratings.service.ts:1]: The current nohat script is just that, a standalone nodejs script that is currently called with a system cron job. It works however, it's kinda a mess there's no comments explaining what it's doing, logging, no unit tests, no automated deployments, or following any standards that we're now defining with the playtak-api.

The goal is is bring it into the playtak api that runs the generateRating with the nestjs schedule rather than the system cron job, including all the other updates in this PR.

lol try reading the nohat version. I'm pretty confident on this now with reviewing the unit testing again and compared generated rating data locally

File: src/models/ratings/ratings.service.spec.ts
Comment (invaderb) [reply src/models/ratings/ratings.service.spec.ts:1]: I spent a bit of time going through the nohat script line by line one what it's doing and replicated that in the unit tests.

File: previous.txt [hunk 17]
```diff
@@ -0,0 +1 @@
+547436 hash
```
Comment (nitzel) [inline previous.txt:1]: Is this equivalent to selecting the oldest game entry in the `games` table that doesn't have ratings yet? Wouldn't that be preferable over using an external file to tell us where to start the rating algorithm?

Comment (invaderb) [reply previous.txt:1]: Currently the Tak server always saves a completed game with a default adjusted rating then the rating script comes through and updates to what the correct adjusted rating is which is difficult to ensure we could use that to allow us to see what games have or haven't been rated. 

In the future we can make a change to the games db to start tracking if a game has actually been rated or not which would allow us to remove the need for saving the last game ID.
